### PR TITLE
mgr/dashboard: fix errors related to frontend service subscriptions.

### PR DESCRIPTION
**Pull request by alfonsomthd**
- Created TimerService: for getting a scheduled observable that checks if there are any observers subscribed before sending a request.
Now FeatureTogglesService uses it internally (solving a hidden bug in it).
- SummaryService & RbdMirroringService: instead of having window.setInterval/setTimeout tasks launched in constructor, now they have a method for polling (and subscribing to it). These methods are called in the appropriate component so we unsubscribe properly.
- Fixed: some subscriptions were not being unsubscribed properly.
- RbdFormComponent: little refactoring and fixed (also improved) unit tests.

Topic to discuss after merge: implement a consensual unsubscription mechanism/guideline that applies to the whole project?

Fixes: https://tracker.ceph.com/issues/44228
Signed-off-by: Alfonso Martínez <almartin@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts [hunk 1]
```diff
@@ -1,46 +1,62 @@
...
+    this.initDataSource();
     this.refresh();
 
     this.ngZone.runOutsideAngular(() => {
-      this.polling = window.setInterval(() => {
+      this.intervalSubscription = interval(this.INTERVAL).subscribe(() => {
         this.ngZone.run(() => {
-          this.refresh();
+          if (this.summaryDataSource.observers.length === 0) {
+            return;
```
[the remark this answers is not in the document]
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts]: Can you point to the example?
Currently I am seeing 401 unauthorized requests:
/api/feature_toggles
after logging out.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts [hunk 2]
```diff
@@ -1,46 +1,62 @@
...
+          } else if (_.includes(['/login', '/login-change-password'], this.router.url)) {
```
[the remark this answers is not in the document]
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts]: We have take into account is not enough to be logged in, butto tkae int oaccount current permissions, as the former user does not have to be the current one.

**Comment by alfonsomthd** [comment]
> The existing implementation of these two services was already in my opinion a bit messy for the thing it should basically do: polling an endpoint. So I wouldn't add more 'concerns' to that (stop polling if something external happens).
> 
> For dealing with the log-out side-effect I'd simply go to the existing HTTP Interceptor logic (or the JWT one) and handle that stuff there.
> 
> For the unsubscribe issue, given it's a general problem not limited to these 2 services, I'd consider adopting some Angular pattern (`AsyncPipe`, rxjs `takeUntil`/`takeWhile`) that deals with unsubscribes or some external library to help on it:
> 
>     * [ngx-auto-unsubscribe](https://www.npmjs.com/package/ngx-auto-unsubscribe)
> 
>     * [ngx-componentdestroyed](https://www.npmjs.com/package/@w11k/ngx-componentdestroyed)

I'll look for a more general solution, but take into account that using things like AsyncPipe they unsubscribe you and you will not re-subscribe again as they are singleton services,
so we have to implement a proper behaviour with subscriptions and not lose the wanted ones in the process.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts [hunk 3]
```diff
@@ -1,46 +1,62 @@
...
+  }
+
   enablePolling() {
+    this.initDataSource();
     this.refresh();
 
     this.ngZone.runOutsideAngular(() => {
-      this.polling = window.setInterval(() => {
+      this.intervalSubscription = interval(this.INTERVAL).subscribe(() => {
         this.ngZone.run(() => {
```
[the remark this answers is not in the document]
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts]: Ack. As there are several intervals running outside angular, we can unify them into a single valid approach.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts]: You're right, but receiving a 401 when you are not logged would be expected behaviour (`Unauthorized`), and different from getting a 403 in the same case (`Forbidden`). We may want to suspend polling while not user credentials are provided. For the 403 scenario, those components using services that require back-end permissions should unsubscribe on destroy.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts]: Yup, but that's an issue caused by the fact that `NavigationComponent` is not ubsubscribing from FeatureTogglesService on `ngOnDestroy` :/

[the remark this answers is not in the document]
Comment (alfonsomthd) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/summary.service.ts]: In fact it's related to this:
https://github.com/ceph/ceph/pull/34058#discussion_r397179685

**Comment by alfonsomthd** [comment]
@epuertat @tspmelo Updated PR according to previous conversation.

**Comment by LenzGr** [comment]
jenkins run make check

**Comment by LenzGr** [comment]
jenkins test make check

**Comment by tspmelo** [comment]
jenkins test make check

**Pull request merged**